### PR TITLE
Rewrite API to use ColorSpace objects with a tree hierarchy

### PR DESCRIPTION
**Pull request by LeaVerou**
This will close a variety of issues including: 

- #113
- #144 
- #145
- Part of #111 
- #112 

It also introduces a variety of nice things such as:

Each color space is derived by another as a "base". These base pointers create a tree with a root of XYZ D65. 
 
image

Converting between color spaces now finds the lowest common ancestor of the two color spaces in the tree, eliminating the need for "shortcuts" so that Color.js does the right thing and doesn't convert e.g. HSV to HSL through XYZ.

Rewrite is close to being done, regressions I still need to fix:
- Gamut mapping seems to be slightly off
- Interpolated colors are a bit off

### CODE REVIEW

File: src/spaces/p3.js [hunk 1]
```diff
@@ -1,23 +1,17 @@
...
 
-Color.defineSpace({
-	inherits: "srgb",
+export default RGBColorSpace.create({
 	id: "p3",
 	name: "P3",
+	base: P3Linear,
 	cssId: "display-p3",
 	// Gamma correction is the same as sRGB
 	// convert an array of display-p3 values to CIE XYZ
```
Comment (svgeesus) [inline src/spaces/p3.js:11]: Due to the matrix moving to p3-linear, all but the first line in this comment can be deleted

**Review by svgeesus — APPROVED** [review]
A few small comments and suggestions, address if you want, but nothing that would block approving

File: src/adapt.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		return XYZ;
+	}
+
+	let env = {W1, W2, XYZ, options};
+
+	hooks.run("chromatic-adaptation-start", env);
+
+	if (!env.M) {
+		if (env.W1 === Color.WHITES.D65 && env.W2 === Color.WHITES.D50) {
+			// Linear Bradford CAT
```
[the remark this answers is not in the document]
Comment (LeaVerou) [reply src/adapt.js:41]: Feel free

File: src/adapt.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+			// ];
+
+			env.M = [
+				[  1.0479298208405488,    0.022946793341019088,  -0.05019222954313557 ],
+				[  0.029627815688159344,  0.990434484573249,     -0.01707382502938514 ],
+				[ -0.009243058152591178,  0.015055144896577895,   0.7518742899580008  ]
+			];
+		}
+		else if (env.W1 === Color.WHITES.D50 && env.W2 === Color.WHITES.D65) {
+			// env.M = [
```
[the remark this answers is not in the document]
Comment (LeaVerou) [reply src/adapt.js:55]: Same :)

File: src/keywords.js [hunk 4]
```diff
@@ -1,16 +1,12 @@
-/* Parse color keywords without the browser DOM
- * This is only needed to parse Color keywords in Node,
- * and to improve performance when parsing color keywords in the browser
- * To take advantage of this, just import the module.
- * You can also take advantage of its default exports, if you need a data structure of named colors
- * Note that this does not handle currentColor
+/* List of CSS color keywords
+ * Note that this does not include currentColor and transparent
```
[the remark this answers is not in the document]
Comment (LeaVerou) [reply src/keywords.js:2]: `transparent` is handled in code: https://github.com/LeaVerou/color.js/blob/colorspace-rewrite/src/spaces/srgb.js#L102=

**Review by svgeesus — APPROVED** [review]

**Pull request merged**